### PR TITLE
feat: Improved endpoint design

### PR DESCRIPTION
**Pull request by elonniu**
## Description

- Using Lambda Runtime instead of Docker Image.
- Fixed Endpoint status loss or out of sync issues.
- Decouple message notifications using EventBridge.
- Use `aws_lambda_powertools`.
- Release new RESTful API for `DELETE` endpoints.
- Validate input parameters using API Gateway Model.
- Formatted changed files.

## Type of change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

### CODE REVIEW

File: middleware_api/lambda/inference/app.py [hunk 1]
```diff
@@ -400,23 +400,24 @@ async def deploy_sagemaker_endpoint(request: Request):
         logger.info(f"input in json format {payload}")
         payload['endpoint_deployment_id'] = endpoint_deployment_id
 
+        # put the item to inference DDB for later check status
+        # must insert item first
+        current_time = str(datetime.now())
```
Comment (NingLu) [inline middleware_api/lambda/inference/app.py:405]: Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The naive datetime objects are treated by many datetime methods as local times, it is preferred to use aware datetimes to represent times in UTC. The recommended way to create an aware datetime object representing a specific timestamp in UTC is by passing `tzinfo` as an argument to the method.

[Learn more](https://docs.python.org/3/library/datetime.html#aware-and-naive-objects)

Similar issue at line number 428.

File: middleware_api/lambda/inference_v2/sagemaker_endpoint_events.py [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+
+def get_endpoint_with_endpoint_name(endpoint_name):
+    try:
+        resp = endpoint_deployment_table.scan(
+            FilterExpression=Attr('endpoint_name').eq(endpoint_name)
+        )
+        logger.info(resp)
+        record_list = resp['Items']
+        if len(record_list) == 0:
```
Comment (NingLu) [inline middleware_api/lambda/inference_v2/sagemaker_endpoint_events.py:53]: Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

To check if a container or sequence (string, list, tuple) is empty, use `if not val`. Do not compare its length using `if len(val) == 0`

[Learn more](https://www.python.org/dev/peps/pep-0008/#programming-recommendations#:~:text=if%20not%20seq)

File: infrastructure/src/sd-inference/sagemaker-endpoints-event.ts [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+        });
+
+        const rule = new Rule(this.scope, 'SageMakerEndpointStateChangeRule', {
+            eventPattern: {
+                source: ['aws.sagemaker'],
+                detailType: ['SageMaker Endpoint State Change'],
+                detail: {
+                    EndpointName: [
+                        {
+                            prefix: 'infer-endpoint-'
```
Comment (alvindaiyan) [inline infrastructure/src/sd-inference/sagemaker-endpoints-event.ts:103]: does we defined this prefix?

**Review by alvindaiyan — CHANGES_REQUESTED** [review]
overall, this is awesome changes

File: infrastructure/src/sd-inference/sagemaker-endpoints-event.ts (hunk 3 above)
Comment (elonniu) [reply infrastructure/src/sd-inference/sagemaker-endpoints-event.ts:103]: yes

**Review by alvindaiyan — APPROVED** [review]
gooooooooo

**Pull request merged**